crates/runtime/src/dataconnector/graphql.rs

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use arrow::array::RecordBatch;
+use arrow::datatypes::{Schema, SchemaRef};
+use arrow_json::reader::infer_json_schema_from_iterator;
+use arrow_json::ReaderBuilder;
+use async_trait::async_trait;
+use datafusion::execution::context::SessionState;
+use datafusion::logical_expr::Expr;
+use datafusion::physical_plan::ExecutionPlan;
+use itertools::Itertools;
+use reqwest::header::{HeaderMap, HeaderValue};
+use reqwest::header::{CONTENT_TYPE, USER_AGENT};
+use secrets::{get_secret_or_param, Secret};
+use serde_json::Value;
+use url::Url;
+
+use crate::component::dataset::Dataset;
+use datafusion::datasource::{MemTable, TableProvider, TableType};
+use snafu::prelude::*;
+use std::any::Any;
+use std::io::Cursor;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::{collections::HashMap, future::Future};
+
+use super::{DataConnector, DataConnectorFactory};
+
+pub struct GraphQL {
+    secret: Option<Secret>,
+    params: Arc<HashMap<String, String>>,
+}
+
+impl DataConnectorFactory for GraphQL {
+    fn create(
+        secret: Option<Secret>,
+        params: Arc<HashMap<String, String>>,
+    ) -> Pin<Box<dyn Future<Output = super::NewDataConnectorResult> + Send>> {
+        Box::pin(async move {
+            let graphql = Self { secret, params };
+            Ok(Arc::new(graphql) as Arc<dyn DataConnector>)
+        })
+    }
+}
+
+enum Auth {
+    Basic(String, Option<String>),
+    Bearer(String),
+}
+
+struct GraphQLClient {
+    client: reqwest::Client,
+    endpoint: Url,
+    query: String,
+    json_path: String,
+    auth: Option<Auth>,
+}
+
+impl GraphQLClient {
+    async fn execute(&self) -> datafusion::error::Result<(Vec<Vec<RecordBatch>>, SchemaRef)> {
+        let body = format!(r#"{{"query": "{}"}}"#, self.query.lines().join(" "));
+        let mut request = self.client.post(self.endpoint.clone()).body(body);
+
+        match &self.auth {
+            Some(Auth::Basic(user, pass)) => {
+                request = request.basic_auth(user, pass.clone());
+            }
+            Some(Auth::Bearer(token)) => {
+                request = request.bearer_auth(token);
+            }
+            _ => {}
+        }
+
+        let response = request.send().await.map_err(|_| {
```

Comment:
Please include internal error, otherwise it will be hard to troubleshoot the problem

Reply:
Graphql has a standard spec for errors https://spec.graphql.org/October2021/#sec-Errors, ideally connector should be able to parse that and map to generic graphql error specified in `DataConnectorError`

Reply:
I think it's better to do in a separate pr

Reply:
And this is not a graphql error, it's reqwest error